acra/configuring-maintaining/optimisations/acrastructs_vs_acrablocks.md

```diff
@@ -0,0 +1,27 @@
+---
+title: AcraStructs vs AcraBlocks
+weight: 3
+---
+
+# AcraStructs vs AcraBlocks
+
+There are two different "crypto containers" AcraServer can use to store encrypted value in.
+Then, these containers are stored in database and decrypted back by AcraServer when needed.
+
+History
+* `AcraStruct` is our "old" one, though we're not going to remove it any time soon.
+* `AcraBlock` on the other hand is new and proposes some extra features.
+
+Encryption with a key
+* `AcraStruct` uses key provided from somewhere outside (by Acra) to encrypt the value.
+* `AcraBlock` uses random key to encrypt the value, then encrypts this key with another key, this time provided from outside.
+  That first key is called DEK (data encryption key) and the second one is called KEK (key encryption key).
+
+Key rotation (when you need re-encrypt data with a new key)
+* With `AcraStruct`, the value has to be decrypted completely, then encrypted with a new key.
```

Comment:
actually AcraStructs uses KEK and DEK too. But for KEK used asymmetric encryption that are slow. And you need to load private key, extract public key from AcraStruct, compute ECDHE to get "shared key" that used as KEK. And after that decrypt DEK.
When for AcraBlock AcraServer uses symmetric key and just decrypts KEK that encrypted and stored in AcraBlock. And decryption uses symmetric encryption that are much faster than asymmetric.

Reply:
Turned out I don't know their internal layout so well. Gonna update based on description from "Data structures" page.